scripts/ui/controlnet_unit.py

```diff
@@ -0,0 +1,921 @@
+import gradio as gr
+import functools
+from typing import List, Optional, Union, Dict, Callable
+import numpy as np
+import base64
+
+from scripts.utils import svgPreprocess, get_module_basename
+from scripts import (
+    global_state,
+    external_code,
+    processor,
+    batch_hijack,
+)
+from scripts.processor import (
+    preprocessor_sliders_config,
+    flag_preprocessor_resolution,
+    model_free_preprocessors,
+    preprocessor_filters,
+    HWC3,
+)
+from modules import shared
+from modules.ui_components import FormRow
+
+
+class ToolButton(gr.Button, gr.components.FormComponent):
+    """Small button with single emoji as text, fits inside gradio forms"""
+
+    def __init__(self, **kwargs):
+        super().__init__(variant="tool", elem_classes=["cnet-toolbutton"], **kwargs)
+
+    def get_block_name(self):
+        return "button"
+
+
+class UiControlNetUnit(external_code.ControlNetUnit):
+    """The data class that stores all states of a ControlNetUnit."""
+
+    def __init__(
+        self,
+        input_mode: batch_hijack.InputMode = batch_hijack.InputMode.SIMPLE,
+        batch_images: Optional[Union[str, List[external_code.InputImage]]] = None,
+        output_dir: str = "",
+        loopback: bool = False,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.is_ui = True
+        self.input_mode = input_mode
+        self.batch_images = batch_images
+        self.output_dir = output_dir
+        self.loopback = loopback
+
+
+def update_json_download_link(json_string: str, file_name: str) -> Dict:
+    base64_encoded_json = base64.b64encode(json_string.encode("utf-8")).decode("utf-8")
+    data_uri = f"data:application/json;base64,{base64_encoded_json}"
+    style = """ 
+    position: absolute;
+    right: var(--size-2);
+    bottom: calc(var(--size-2) * 4);
+    font-size: x-small;
+    font-weight: bold;
+    padding: 2px;
+
+    box-shadow: var(--shadow-drop);
+    border: 1px solid var(--button-secondary-border-color);
+    border-radius: var(--radius-sm);
+    background: var(--background-fill-primary);
+    height: var(--size-5);
+    color: var(--block-label-text-color);
+    """
+    hint = "Download the pose as .json file"
+    html = f"""<a href='{data_uri}' download='{file_name}' style="{style}" title="{hint}">
+                Json</a>"""
+    return gr.update(value=html, visible=(json_string != ""))
+
+
+class ControlNetUnit(object):
```

Comment:
consider renaming to `UiControlNetUnit`, and rename the current `UiControlNetUnit` to `BatchControlNetUnit` or something like that. Having 2 classes named `ControlNetUnit` in the repo is going to make it harder to understand the code.

Reply:
Maybe `InternalControlNetUnit`? To indicate it is used by the extension but not intended for use by the APIs.

Reply:
I think that's fine as well.